flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/wmassigners/WatermarkAssignerOperator.java

```diff
@@ -162,13 +172,14 @@ public void onProcessingTime(long timestamp) throws Exception {
         }
 
         if (processedElements != lastIdleCheckProcessedElements) {
-            timeSinceLastIdleCheck = now;
+            timeSinceLastIdleCheck = inputActivityClock.relativeTimeMillis();
             lastIdleCheckProcessedElements = processedElements;
         }
 
         if (isIdlenessEnabled()
                 && currentStatus.equals(WatermarkStatus.ACTIVE)
-                && timeSinceLastIdleCheck + idleTimeout <= now) {
+                && timeSinceLastIdleCheck + idleTimeout
+                        <= inputActivityClock.relativeTimeMillis()) {
```

Comment:
Extract `inputActivityClock.relativeTimeMillis()` into a local variable to avoid one sys call and `synchronized`?

Reply:
Not that it matters when called that rarely but sure :)